File: programmers/2019KAKAO/2.py
def checkBracket(str):
    stack = []
    for i in str:
        if i == "(":
            stack.append("(")
        else:
            if len(stack) != 0:
                stack.pop()
            else:
                return False

    return True if len(stack) == 0 else False


def makeUV(str):
    result = ""
    rU, rV = "", ""

    if str == "":
        return str

    for i, string in enumerate(str):
        rU += string
        if rU.count("(") > 0 and rU.count("(") == rU.count(")"):
            rV = str[i + 1 :]
            break

    if checkBracket(rU):
        string = makeUV(rV)
        if string == "":
            return rU
        else:
            result = rU + string
            return result

    else:
        v = makeUV(rV)
        u = ""
        tmp_u = rU[1 : len(rU) - 1]
        for i in range(0, len(tmp_u)):
            if tmp_u[i] == "(":
                u += ")"
            else:
                u += "("
        result = "(" + v + ")" + u
        return result


def solution(p):
    answer = ""
    if checkBracket(p):
        answer = p
    else:
        answer = makeUV(p)

    print(answer)
    return answer


input_p = "()))((()"
# "))))(((("
solution(input_p)

File: 2019/programmers/2019KAKAO/1/2.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>
using namespace std;

bool checkBracket(string str)
{
    stack<char> s;
    int len = (int)str.length();
    for (int i = 0; i < len; i++)
    {
        char c = str[i];
        if (c == '(')
            s.push(str[i]);

        else
        {
            if (!s.empty())
                s.pop();
            else
                return false;
        }
    }
    return s.empty();
}

string makeUV(string str)
{
    string u = "", v = "";
    int open = 0, close = 0;
    if (str == "")
        return str;

    for (int i = 0; i < str.size(); i++)
    {
        u += str[i];
        open = count(u.begin(), u.end(), '(');
        close = count(u.begin(), u.end(), ')');
        if (u.size() > 0 && open == close)
        {
            v = str.substr(i + 1);
            break;
        }
    }
    if (checkBracket(u))
    {
        string result = makeUV(v);
        if (result == "")
            return u;
        else
        {
            cout << u + result << '\n';
            return u + result;
        }
    }
    else
    {
        string result = "";
        v = makeUV(v);
        string tmp_u = str.substr(1, str.size() - 2);
        u = "";
        for (int i = 0; i < u.size() / 2; i++)
            u += "()";
        result = "(" + u + ")" + v;
        return result;
    }
}
string solution(string p)
{
    string answer = "";
    if (checkBracket(p))
        answer = p;
    else
    {
        makeUV(p);
    }
    return answer;
}

int main(void)
{
    string p = "()))((()";
    cout << solution(p) << '\n';
    return 0;
}